pynear/include/BuiltinSerializers.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <Eigen/Core>
+#include <vector>
+
+namespace vptree {
+
+// Built in serializer for vector of vectors
+template <typename T> void ndarraySerializer(const std::vector<std::vector<T>> &input, std::vector<uint8_t> &output) {
+    /*
+     * This serializer function should write state from lower to higher addresses.
+     * The serializer must append to the end of the file, like in a bytestream.
+     * It writes a ndarray like vector (e.g: a vector of vectors) to a byte array.
+     * It will serialize meta information about the vector such as its size and dimension.
+     * The final format will be:
+     * [totalSize(size_t)][dimension(size_t)][element1(std::vector<T>)]...[elementN(std::vector<T>)]
+     *
+     * See SerializableVPTree declaration for more information.
+     */
+    auto totalSize = input.size();
+    if (totalSize == 0) {
+        return;
+    }
+
+    size_t dimension = input[0].size();
+    auto totalBytes = totalSize * dimension * sizeof(T);
+    // add space for all elements of given dimensions + total size + dimension in bytes
+    size_t startPoint = output.size();
+    output.resize(output.size() + totalBytes + 2 * sizeof(size_t));
+
+    uint8_t *data = output.data() + startPoint;
+    // store total size and dimension first
+    (*(size_t *)(data)) = totalSize;
+    data += sizeof(size_t);
+    (*(size_t *)(data)) = dimension;
+    data += sizeof(size_t);
+
+    // store elements
+    for (const auto &element : input) {
+        std::memcpy(data, &element.front(), dimension * sizeof(T));
+        data += dimension * sizeof(T);
+    }
+};
+
+// Built in deserializer for vector of vectors
+template <typename T> std::vector<std::vector<T>> ndarrayDeserializer(const uint8_t *input, size_t &readBytes) {
```

Comment:
the serializer and deserilizer should have symmetric signatures I think
the serializer gets a uint vector ref, while this one takes a pointer
maybe they should use the same

Reply:
We can have that. But this is somehow to reinforce that one needs to write so users have container support. The other is for reading, so users dont need (and actually cannot) modify the buffer. It is somehow used strategy to pass a const pointer for reading and container for writting.